### PR TITLE
mtsLora. add mtx lora codec

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

File: src/mtxLora/constants/commandRelations.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+export const responseByRequest = new Map();
+
+// id to constructor
+export const requestById = new Map(Object.values(downlinkCommands).map(command => [command.id, command]));
+export const responseById = new Map(Object.values(uplinkCommands).map(command => [command.id, command]));
+
+
+requestByResponse.set(uplinkCommands.GetHalfhoursEnergies, downlinkCommands.GetHalfhoursEnergies);
+
+
```
Comment (DarkPark) [inline src/mtxLora/constants/commandRelations.ts:18]:
```suggestion
```

File: src/mtxLora/commands/uplink/GetDayMaxPower.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    toBytes (): Uint8Array {
+        const {parameters: {date, powers}} = this;
+        const buffer = new CommandBinaryBuffer(COMMAND_HEADER_SIZE + COMMAND_MAX_SIZE);
+
+        // body
+        buffer.setDate(date);
+        buffer.setTariffsPowerMax(powers);
+
+        const header = new Uint8Array([
+            COMMAND_ID,
```
Comment (DarkPark) [inline src/mtxLora/commands/uplink/GetDayMaxPower.ts:111]:
```suggestion
        return new Uint8Array([
            COMMAND_ID,
            buffer.position,
            ...buffer.getBytesToOffset()
        ]);
```

?

File: src/mtxLora/commands/uplink/GetDayEnergies.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import mergeUint8Arrays from '../../../utils/mergeUint8Arrays.js';
+import {UPLINK} from '../../../constants/directions.js';
+
+
+const COMMAND_ID = 0x78;
+const DATE_SIZE = 3; // year, month, day
+const MAX_TARIFFS_ENERGIES_SIZE = 5 * 4 * 4; // 5 energy types, 4 tariffs, 4 bytes - energy value
+const COMMAND_MAX_SIZE = DATE_SIZE + MAX_TARIFFS_ENERGIES_SIZE;
+
+interface IGetDayEnergies {
```
Comment (DarkPark) [inline src/mtxLora/commands/uplink/GetDayEnergies.ts:12]: better place interface block before const

in other files as well

File: src/mtxLora/commands/uplink/GetDayMaxPower.ts (hunk 2 above)
Comment (DarkPark) [reply src/mtxLora/commands/uplink/GetDayMaxPower.ts:111]: in other files as well

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    const booleanObject = Object.fromEntries(Object.entries(energies).map(([name, value]) => [name, !!value]));
+
+    return bitSet.fromObject(energiesMask, (booleanObject as unknown) as bitSet.TBooleanObject);
+};
+
+type THalfhoursEnergy = Array<number | undefined>;
+
+export type THalfhoursEnergies = IEnergies<THalfhoursEnergy>;
+
+const UndefinedEnergyValue = 0xffffffff;
```
Comment (DarkPark) [inline src/mtxLora/CommandBinaryBuffer.ts:41]:
```suggestion
const undefinedEnergyValue = 0xffffffff;
```

?
or maybe in UPPER_CASE

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            getEnergiesFlags(energies)
+        );
+    }
+
+    getHalfhoursEnergy ( numberOfHalfhours: number ): THalfhoursEnergy {
+        const halfhours = [];
+        let value;
+
+        for ( let index = 0; index < numberOfHalfhours; index++ ) {
+            value = this.getUint32();
```
Comment (DarkPark) [inline src/mtxLora/CommandBinaryBuffer.ts:121]:
```suggestion
            const value = this.getUint32();
```
?

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+        return halfhours;
+    }
+
+    setHalfhoursEnergy ( halfhours: THalfhoursEnergy | undefined ) {
+        let value;
+
+        if ( halfhours ) {
+            for ( let index = 0; index < halfhours.length; index++ ) {
+                value = halfhours[index];
```
Comment (DarkPark) [inline src/mtxLora/CommandBinaryBuffer.ts:134]:
```suggestion
                const value = halfhours[index];
```
?

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    setHalfhoursEnergies ( energies: THalfhoursEnergies ) {
+        this.setHalfhoursEnergy(energies.aPlus);
+        this.setHalfhoursEnergy(energies.aPlusRPlus);
+        this.setHalfhoursEnergy(energies.aPlusRMinus);
+        this.setHalfhoursEnergy(energies.aMinus);
+        this.setHalfhoursEnergy(energies.aMinusRPlus);
+        this.setHalfhoursEnergy(energies.aMinusRMinus);
+    }
+
+    getAPlusTariffEnergies ( energyFlags: number ) : IEnergies {
```
Comment (DarkPark) [inline src/mtxLora/CommandBinaryBuffer.ts:180]:
```suggestion
    getAPlusTariffEnergies ( energyFlags: number ): IEnergies {
```

and in other places

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        }
+
+        return tariffs;
+    }
+
+    setTariffsPowerMax ( tariffs: TTariffsPowerMax ) {
+        let energiesFlags = 0;
+        let tariffsFlags = 0;
+
+        tariffs.forEach( (tariff, index) => {
```
Comment (DarkPark) [inline src/mtxLora/CommandBinaryBuffer.ts:380]:
```suggestion
        tariffs.forEach((tariff, index) => {
```

File: src/mtxLora/constants/commandRelations.ts (hunk 1 above)
Comment (Pnom) [reply src/mtxLora/constants/commandRelations.ts]: fixed

File: src/mtxLora/commands/uplink/GetDayMaxPower.ts (hunk 2 above)
Comment (Pnom) [reply src/mtxLora/commands/uplink/GetDayMaxPower.ts]: fixed

File: src/mtxLora/commands/uplink/GetDayEnergies.ts (hunk 3 above)
Comment (Pnom) [reply src/mtxLora/commands/uplink/GetDayEnergies.ts:6]: fixed

File: src/mtxLora/CommandBinaryBuffer.ts (hunk 4 above)
Comment (Pnom) [reply src/mtxLora/CommandBinaryBuffer.ts]: fixed

File: src/mtxLora/CommandBinaryBuffer.ts [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+ * Command specific byte array manipulation.
+ */
+class CommandBinaryBuffer extends BinaryBuffer {
+    constructor ( dataOrLength: Uint8Array | number | string ) {
+        // force BE for all numbers
+        super(dataOrLength, false);
+    }
+
+    getDate (): IDate {
+        const date1 = this.getUint8();
```
[the remark this answers is not in the document]
Comment (Pnom) [reply src/mtxLora/CommandBinaryBuffer.ts]: fixed

File: src/mtxLora/CommandBinaryBuffer.ts (hunk 5 above)
Comment (Pnom) [reply src/mtxLora/CommandBinaryBuffer.ts]: fixed

File: src/mtxLora/CommandBinaryBuffer.ts (hunk 7 above)
Comment (Pnom) [reply src/mtxLora/CommandBinaryBuffer.ts]: fixed

File: src/mtxLora/CommandBinaryBuffer.ts (hunk 8 above)
Comment (Pnom) [reply src/mtxLora/CommandBinaryBuffer.ts:379]: fixed

File: src/mtxLora/commands/uplink/GetHalfhoursEnergies.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        buffer.setUint8(firstHalfhour);
+        buffer.setUint8(numberOfHalfhours);
+        buffer.setHalfhoursEnergies(energies);
+
+        const header = new Uint8Array([
+            COMMAND_ID,
+            buffer.position
+        ]);
+
+        return mergeUint8Arrays(header, buffer.getBytesToOffset());
```
Comment (DarkPark) [inline src/mtxLora/commands/uplink/GetHalfhoursEnergies.ts:119]:
```suggestion
```

Comment (Pnom) [reply src/mtxLora/commands/uplink/GetHalfhoursEnergies.ts]: fixed

**Review by DarkPark — DISMISSED** [review]

**Review by DarkPark — DISMISSED** [review]

File: src/mtxLora/commands/uplink/GetDayMaxPower.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ * console.log(command.parameters);
+ * // output:
+ * {
+ *     date: {
+ *         year: 21,
+ *         month: 2,
+ *         day: 3
+ *     },
+ *     energies: [
```
Comment (DeadByElpy) [inline src/mtxLora/commands/uplink/GetDayMaxPower.ts:67]: powers?

**Review by DeadByElpy — CHANGES_REQUESTED** [review]

**Review by DeadByElpy — APPROVED** [review]

File: src/mtxLora/commands/uplink/GetDayMaxPower.ts (hunk 11 above)
Comment (Pnom) [reply src/mtxLora/commands/uplink/GetDayMaxPower.ts]: fixed

**Pull request merged**